internal/speakerlist/speakerlist.go

```diff
@@ -80,36 +83,107 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
 }
 
 // Start starts the needed goroutines
-func (sl *SpeakerList) Start(client *k8s.Client) error {
+func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.client = client
 
 	if sl.mList == nil {
-		return nil
+		return
 	}
 
 	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
 
-	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
-	if err != nil {
-		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
-		return err
+func (sl *SpeakerList) memberlistJoin() {
+	if sl.mList == nil {
+		return
+	}
+	joinList := sl.joinList()
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) joinList() []string {
+	joinList := []string{}
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return joinList
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	for name, ip := range sl.k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	return joinList
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.memberlistJoin()
```

Comment:
Is the `memberlist.Join()` something light if all nodes are already in the cluster, to be running it every 10 secs?

Reply:
`joinList()` makes the diff between `k8sSpeakers` and `mlSpeakers` to be sure we only try to join the nodes that are not joined

Reply:
Right. Sorry for the noise! :)